### ISSUE TITLE
support disable ipv6 candidate in RTCConfiguration

### ISSUE THREAD

**Issue by bdrtc** [w3c/webrtc-extensions#138]
Ipv6 is enabled by default in some browsers and there are ipv6 candidates generated by default during setup peerconnection call, 
but not all webrtc based system support ipv6 now, ipv6 candidates is not nessesary in this case,
What's more,  ipv6 candidates in sdp may cause negotiation error when interoperation with some rtc servers which does't support ipv6 address resolution.
Disable ipv6 will reduce the time cost of candidates collection especially when multiple NICs exist.

I sugguest add an IPv6 configuration parameter like below :
`new RTCPeerConnection({}, {mandatory:{IPv6:false}});`

[chrome support this](https://chromestatus.com/feature/5184046454800384) via googIPv6 ever but remove it as it's not standard API.

**Comment by henbos** [comment]
That's right, as of M108 it is not possible to disable IPv6 using googIPv6.

I'm not necessarily a huge fan of adding it back (are there other work-arounds?), but if we did, it should be part of the RTCConfiguration (first parameter), not the legacy constraints dictionary (second parameter) since that has been unshipped and I don't see the point of more parameters or this "mandatory" syntax which is also legacy and serves no purpose here

**Comment by fippo** [comment]
You can still not signal those candidates - if neither side does there will be no attempts to pair IPv6.

**Comment by bdrtc** [comment]
if we remove it because of it's non-standard API then we can standarded it, 
yeah, even we have other workaround method to diabled it, it's still not standard and most developers maybe do not know it.
As developer we may first check peerconnection configuration parameters.

**Comment by fippo** [comment]
Historically this API was used to allow opt-in to IPv6 support when it was first added back in 2014, see https://bugs.chromium.org/p/webrtc/issues/detail?id=1406
It was never designed to opt-out.

**Comment by steely-glint** [comment]
The right place to filter ipv6 candidates is at the receiving end - the sender can't know if ipv6 candidates might be useful - the recipient can. 

@bdrtc do you have any timings on the delay caused by creating ipv6 candidates ?

**Comment by henbos** [comment]
> if we remove it because of it's non-standard API then we can standarded it,

Agree but IPv6 is here to stay so disabling them is not a very compelling use case X years from now if there are ways to work around the issue already.

**Comment by bdrtc** [comment]
> The right place to filter ipv6 candidates is at the receiving end - the sender can't know if ipv6 candidates might be useful - the recipient can.
> 
> @bdrtc do you have any timings on the delay caused by creating ipv6 candidates ?

when we use full ice case, offer sdp will wait for full ice gather end then send out ,  disable not used NICs/Transport type/Address type will speed up the candidate collection time.

i ever test an old sip server support webrtc, but it dose not support ipv6 candidate exist in sdp which cause sdp negotiation failed, and resolved it by disabled ipv6 candidate collection via googIPv6,  but it's removed after 108, from the descrition of remove resean is its private opt.

 then, open it here for discussion, maybe other users have same  requirment.

**Comment by henbos** [comment]
It should be possible to filter out the candidates from the SDP as well.

Regarding collection time of candidates, I think this is fast in the lower layer, but if there are a lot of candidates (i.e. BUNDLE is not used and there are a lot of m= sections) then there can be an artificial delay added by the fact that each candidate gets surfaced in a separate task execution cycle and each candidate fires an event for each m= section. So if there are 100 candidates that should be exposed to JS that delays both candidates and the resolution of the SLD/SRD promise by 100+ task execution cycles, something which adds noticeable delays (we could talk seconds) of "processing time". This is an optimization bug though that could be solved by the browser implementation.

**Comment by bdrtc** [comment]
user can remove by modify sdp, but can't disable  collection step of lower layer :>
Has't seen ipv6 used in  some webrtc based production system, maybe it will be widdly used in future.

**Comment by henbos** [comment]
Why do you care if collection happens at lower layer?

**Comment by steely-glint** [comment]
I'd like to see numbers on the delay caused by IPV6 collection because : 
a) I expect almost all IPV6 host addresses are collected quickly (no stun rtt needed)
b) Any delay is avoidable by using trickle ice.
c) I'd expect the IPV4 TURN TLS candidate to arrive last - so you'll have to wait for that anyway - (this is the works-everywhere fallback candidate)

I do agree that it is sometimes useful to filter the received candidates to those that are likely to succeed - but that really is best done at the receiving end.
